### PR TITLE
Add semver checks

### PR DESCRIPTION
**Pull request by gcemaj**
Use semver for version checks instead of strings

Closes: #132

### CODE REVIEW

File: build.go [hunk 1]
```diff
@@ -170,8 +171,14 @@ func Build(builder Builder, options ...Option) {
...
 
-	API := strings.TrimSpace(ctx.Buildpack.API)
-	if API != "0.5" && API != "0.6" && API != "0.7" {
+	API, err := semver.NewVersion(ctx.Buildpack.API)
+	if err != nil {
+		config.exitHandler.Error(errors.New("version cannot be parsed"))
+		return
+	}
+
+	compatVersionCheck, _ := semver.NewConstraint(">= 0.5, <=0.7")
```
Comment (samj1912) [inline build.go:180]: Could we factor these out as minversion and maxversion and use those as constants here and in the warning below?

File: build.go [hunk 2]
```diff
@@ -279,7 +286,8 @@ func Build(builder Builder, options ...Option) {
 		file = filepath.Join(ctx.Layers.Path, fmt.Sprintf("%s.toml", layer.Name))
 		logger.Debugf("Writing layer metadata: %s <= %+v", file, layer)
 		var toWrite interface{} = layer
-		if API == "0.5" {
+		isV5Check, _ := semver.NewConstraint("= 0.5")
+		if isV5Check.Check(API) {
```
Comment (samj1912) [inline build.go:290]: We can use API.Equals here?

File: build.go [hunk 3]
```diff
@@ -338,7 +347,8 @@ func Build(builder Builder, options ...Option) {
 		file = filepath.Join(ctx.Layers.Path, "launch.toml")
 		logger.Debugf("Writing application metadata: %s <= %+v", file, launch)
 
-		if API == "0.5" {
+		supportDefaultProcessCheck, _ := semver.NewConstraint("> 0.5")
+		if !supportDefaultProcessCheck.Check(API) {
```
Comment (samj1912) [inline build.go:351]: We could use the semver.Greater api here?

File: build.go [hunk 4]
```diff
@@ -89,6 +90,12 @@ type BOM struct {
 	Entries []BOMEntry
 }
 
+// COnstants to track minimum and maximum version
```
Comment (samj1912) [inline build.go:93]:
```suggestion
// Constants to track minimum and maximum supported Buildpack API versions
```

File: build.go [hunk 5]
```diff
@@ -89,6 +90,12 @@ type BOM struct {
...
+const (
+	minBpVersion = "0.5"
+	maxBpVersion = "0.7"
```
Comment (samj1912) [inline build.go:96]:
```suggestion
	MinSupportedBPVersion = "0.5"
	MaxSupportedBPVersion = "0.7"
```

File: build.go [hunk 6]
```diff
@@ -307,7 +320,7 @@ func Build(builder Builder, options ...Option) {
 		}
 	}
 
-	if API != "0.5" && API != "0.6" {
+	if API.GreaterThan(semver.MustParse("0.7")) || API.Equal(semver.MustParse("0.7")) {
```
[the remark this answers is not in the document]
Comment (gcemaj) [reply build.go:323]: isnt 0.6.1 greater than 0.6?  i thought this feature was added from 0.7 onward

[the remark this answers is not in the document]
Comment (samj1912) [reply build.go:323]: Buildpack API versions cannot have patch versions https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpack-api-version

[the remark this answers is not in the document]
Comment (samj1912) [reply build.go:323]: I think it's fine to leave it this way though. The other alternative is to use Compare and check if it's greater than or equal to 0.

**Comment by samj1912** [comment]
@gcemaj could you rebase and update the PR?

File: build.go [hunk 7]
```diff
@@ -89,6 +90,12 @@ type BOM struct {
 	Entries []BOMEntry
 }
 
+// Constants to track minimum and maximum supported Buildpack API versions
+const (
+	MinSupportedBPVersion = "0.5"
```
Comment (samj1912) [inline build.go:95]: Could we also add docstrings for these constants?

File: build.go [hunk 8]
```diff
@@ -89,6 +90,15 @@ type BOM struct {
 	Entries []BOMEntry
 }
 
+// Constants to track minimum and maximum supported Buildpack API versions
+const (
+	// MinSupportedBPVersion minium supported version of the Buildpacks API
```
Comment (samj1912) [inline build.go:95]:
```suggestion
	// MinSupportedBPVersion indicates the minimum supported version of the Buildpacks API
```

File: build.go [hunk 9]
```diff
@@ -89,6 +90,15 @@ type BOM struct {
...
+	MinSupportedBPVersion = "0.5"
+
+	// MaxSupportedBPVersion maximum supported version of the Buildpacks API
```
Comment (samj1912) [inline build.go:98]:
```suggestion
	// MaxSupportedBPVersion indicates the maximum supported version of the Buildpacks API
```

**Review by samj1912 — APPROVED** [review]

**Pull request merged**